scripts/utils.sh

```diff
@@ -0,0 +1,132 @@
+#!/usr/bin/env bash
+
+# Copyright 2019 Google Inc.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#      http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+# Shared functions for deploying solutions
+
+# Get the absolute path to a file
+# Arguments:
+#   $1 - relative path to convert
+# Returns:
+#   Absolute path of file
+function abspath() {
+    if [ -d "$1" ]; then
+        # dir
+        (cd "$1"; pwd)
+    elif [ -f "$1" ]; then
+        # file
+        if [[ $1 = /* ]]; then
+            echo "$1"
+        elif [[ $1 == */* ]]; then
+            echo "$(cd "${1%/*}"; pwd)/${1##*/}"
+        else
+            echo "$(pwd)/$1"
+        fi
+    fi
+}
+
+
+PROJECT_DIR=$(dirname $(abspath $0))
+UTIL_DIR=$(dirname $(abspath $BASH_SOURCE))
+ROOT_DIR=$(dirname $UTIL_DIR)
+
+if ! [ -x "$(command -v gcloud)" ]; then
+    echo 'Error: gcloud is not installed.' >&2
+    exit 1
+fi
+
+if ! [ -x "$(command -v python3)" ]; then
+    echo 'Error: Python not available.' >&2
+    exit 1
+fi
+
+if ! [ -x "$(command -v npx)" ]; then
+    echo 'Error: npx not available.' >&2
+    exit 1
+fi
+
+if [ -z "${GOOGLE_APPLICATION_CREDENTIALS}" ]; then
```

Comment:
Is the idea that the user would run `auth.py`, then set this variable in their shell? Could we instead just have a standard location for the file and then set the variable automatically?

Reply:
Good idea.

Reply:
done